android/titanium/src/java/org/appcelerator/titanium/proxy/TiViewProxy.java

```diff
@@ -916,14 +918,40 @@ protected void handleFocus()
 	}
 
 	@Kroll.method
-	public TiBlob toImage()
+	public TiBlob toImage(final @Kroll.argument(optional=true) KrollFunction callback)
 	{
-		if (TiApplication.isUIThread()) {
-			return handleToImage();
-
-		} else {
-			return (TiBlob) TiMessenger.sendBlockingMainMessage(getMainHandler().obtainMessage(MSG_TOIMAGE), getActivity());
+		final boolean waitForFinish = (callback == null);
+		TiBlob blob;
+		
+		/*
+		 * Callback don't exist. Just render on main thread and return blob. 
+		 */
+		if (waitForFinish) {
+			if (TiApplication.isUIThread()) {
+				blob = handleToImage();
+			} else {
+				blob = (TiBlob) TiMessenger.sendBlockingMainMessage(getMainHandler().obtainMessage(MSG_TOIMAGE), getActivity());
+			}
+		} 
+		
+		/*
+		 * Callback exists. Perform async rendering and return an empty blob.
+		 */
+		else {
+			// Create a non-null empty blob to return. 
+			blob  = TiBlob.blobFromImage(Bitmap.createBitmap(1, 1, Config.ARGB_8888));
+			Runnable renderRunnable = new Runnable() {
```

Comment:
Calling TiMessenger.sendBlockingMainMessage() is performing the msg synchronously while blocking on the main thread. This essentially makes the method execution synchronous. But according to doc and iOS implementation - if a callback is passed, the method is supposed to execute asynchronously. So Im using a runnable to render and send the result through callback. 

Am I missing anything here ?